synapse/handlers/stats.py

```diff
@@ -52,4 +59,324 @@ def __init__(self, hs):
+            except:  # noqa: E722 – re-raised so fine
+                lock.release()
+                raise
+
+    @defer.inlineCallbacks
+    def _unsafe_process(self):
+        # If self.pos is None then means we haven't fetched it from DB
+        if self.pos is None or None in self.pos.values():
+            self.pos = yield self.store.get_stats_positions()
+
+        # If still None then the initial background update hasn't started yet
+        if self.pos is None or None in self.pos.values():
+            return None
+
+        # Loop round handling deltas until we're up to date
+        with Measure(self.clock, "stats_delta"):
+            while True:
+                deltas = yield self.store.get_current_state_deltas(
+                    self.pos["state_delta_stream_id"]
+                )
+                if not deltas:
+                    break
+
+                logger.debug("Handling %d state deltas", len(deltas))
+                yield self._handle_deltas(deltas)
+
+                self.pos["state_delta_stream_id"] = deltas[-1]["stream_id"]
+
+                event_processing_positions.labels("stats").set(
+                    self.pos["state_delta_stream_id"]
+                )
+
+                if self.pos is not None:
+                    yield self.store.update_stats_positions(self.pos)
+
+    @defer.inlineCallbacks
+    def _handle_deltas(self, deltas):
+        """
+        Called with the state deltas to process
+        """
+        for delta in deltas:
+            typ = delta["type"]
+            state_key = delta["state_key"]
+            room_id = delta["room_id"]
+            event_id = delta["event_id"]
+            stream_id = delta["stream_id"]
+            prev_event_id = delta["prev_event_id"]
+            stream_pos = delta["stream_id"]
+
+            logger.debug("Handling: %r %r, %s", typ, state_key, event_id)
+
+            token = yield self.store.get_earliest_token_for_stats("room", room_id)
+
+            # If the earliest token to begin from is larger than our current
+            # stream ID, skip processing this delta.
+            if token is not None and token >= stream_id:
+                logger.debug(
+                    "Ignoring: %s as earlier than this room's initial ingestion event",
+                    event_id,
+                )
+                continue
+
+            if event_id is None and prev_event_id is None:
+                # Errr...
+                continue
+
+            event_content = {}
+
+            if event_id is not None:
+                event = yield self.store.get_event(event_id, allow_none=True)
+                if event:
+                    event_content = event.content or {}
+
+            # We use stream_pos here rather than fetch by event_id as event_id
+            # may be None
+            now = yield self.store.get_received_ts_by_stream_pos(stream_pos)
+            now = int(now) // 1000
```

Comment:
Can we work in milliseconds please.

Reply:
To what extent? Everywhere? Note that `self.stats_bucket_size` is in sec; should I update the (config parser?) to match millisec?

Reply:
Everywhere please :)